src/Sparse/Matrix.jl

```diff
@@ -723,6 +723,98 @@ function *(A::SMat, b)
   return A * base_ring(A)(b)
 end
 
+################################################################################
+#
+#  Dot product
+#
+################################################################################
+
+@doc raw"""
+    dot(x::SRow{T}, A::SMat{T}, y::SRow{T}) where T -> T
+
+Return the generalized dot product `dot(x, A*y)`.
+"""
+function dot(x::SRow{T}, A::SMat{T}, y::SRow{T}) where T
+  v = T(0)
+  px = 1
+  for i in 1:length(A.rows)
+    while px <= length(x.pos) && x.pos[px] < i
+      px += 1
+    end
+    if px > length(x.pos)
+      break
+    elseif x.pos[px] > i
+      continue
+    end
+
+    s = T(0)
+    py = 1
+    for j in 1:length(A[i].pos)
+      while py <= length(y.pos) && y.pos[py] < A[i].pos[j]
+        py += 1
+      end
+      if py > length(y.pos)
+        break
+      elseif y.pos[py] > A[i].pos[j]
+        continue
+      end
+
+      s += A[i].values[j] * y.values[py]
+    end
+
+    v += x.values[px] * s
+  end
+
+  return v
+end
+
+@doc raw"""
+    dot(x::AbstractVector{T}, A::SMat{T}, y::AbstractVector{T}) where T -> T
+
+Return the generalized dot product `dot(x, A*y)`.
+"""
+function dot(x::AbstractVector{T}, A::SMat{T}, y::AbstractVector{T}) where T
+  @req length(x) == length(A.rows) == length(y) "incompatible matrix dimensions"
+
+  v = T(0)
+  for i in 1:length(A.rows)
+    s = T(0)
+    for j in 1:length(A[i].pos)
+      if A[i].pos[j] > length(y)
+        error("incompatible matrix dimensions")
```

Comment:
Is there a reason we do not check this at the start?

Reply:
Because `A` is sparse, we would can't avoid looping over all rows to check this. So I thought we might as well check this on the fly with the assumption that we won't error.

Reply:
Never mind looking at the some code it looks like `A.c` stores this ? If I read this right, I will change it.